File: itx-go/itx-control-worker/internal/dispatcher/dispatcher.go
package dispatcher

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"

	"github.com/chehsunliu/itx/itx-go/itx-contract/email"
	"github.com/chehsunliu/itx/itx-go/itx-contract/queue/message"
	"github.com/chehsunliu/itx/itx-go/itx-contract/repo/post"
	"github.com/chehsunliu/itx/itx-go/itx-control-worker/internal/state"
)

type Dispatcher struct {
	state state.WorkerState
}

func New(state state.WorkerState) *Dispatcher {
	return &Dispatcher{state: state}
}

func (d *Dispatcher) Handle(ctx context.Context, body string) error {
	var env struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal([]byte(body), &env); err != nil {
		return fmt.Errorf("invalid message: %w", err)
	}
	switch env.Type {
	case message.TypePostCreated:
		var msg message.PostCreatedMessage
		if err := json.Unmarshal([]byte(body), &msg); err != nil {
			return fmt.Errorf("invalid post.created: %w", err)
		}
		return d.handlePostCreated(ctx, msg)
	default:
		return fmt.Errorf("unknown message type: %s", env.Type)
	}
}

func (d *Dispatcher) handlePostCreated(ctx context.Context, msg message.PostCreatedMessage) error {
	p, err := d.state.PostRepo.Get(ctx, post.GetParams{ID: msg.PostID})
	if err != nil {
		return err
	}
	author, err := d.state.UserRepo.Get(ctx, msg.AuthorID)
	if err != nil {
		return err
	}
	subscribers, err := d.state.SubscriptionRepo.ListSubscribers(ctx, msg.AuthorID)
	if err != nil {
		return err
	}
	slog.Info(
		"sending post.created notifications",
		"post_id", msg.PostID,
		"author", author.Email,
		"subscribers", len(subscribers),
	)

	for _, sub := range subscribers {
		if err := d.state.EmailClient.Send(ctx, email.SendEmailMessage{
			To:      sub.Email,
			Subject: fmt.Sprintf("%s just published a new post", author.Email),
			Body:    fmt.Sprintf("Check out the new post: %s", p.Title),
		}); err != nil {
			return err
		}
	}

	return d.state.PostRepo.MarkNotified(ctx, msg.PostID)
}

File: itx-rs/itx-control-worker/src/dispatcher.rs
use async_trait::async_trait;
use itx_contract::email::SendEmailMessage;
use itx_contract::queue::message::{MessageBody, PostCreatedMessageBody};
use itx_contract::queue::{HandlerError, MessageHandler};
use itx_contract::repo::post::GetParams;

use crate::state::ControlWorkerState;

pub struct ControlDispatcher {
    state: ControlWorkerState,
}

impl ControlDispatcher {
    pub fn new(state: ControlWorkerState) -> Self {
        Self { state }
    }

    async fn handle_post_created(&self, body: PostCreatedMessageBody) -> Result<(), HandlerError> {
        let post = self.state.post_repo.get(GetParams { id: body.post_id }).await?;
        let author = self.state.user_repo.get(body.author_id).await?;
        let subscribers = self.state.subscription_repo.list_subscribers(body.author_id).await?;
        tracing::info!(
            post_id = body.post_id,
            author = %author.email,
            subscribers = subscribers.len(),
            "sending post.created notifications"
        );

        for subscriber in subscribers {
            self.state
                .email_client
                .send(SendEmailMessage {
                    to: subscriber.email,
                    subject: format!("{} just published a new post", author.email),
                    body: format!("Check out the new post: {}", post.title),
                })
                .await?;
        }

        self.state.post_repo.mark_notified(body.post_id).await?;
        Ok(())
    }
}

#[async_trait]
impl MessageHandler for ControlDispatcher {
    async fn handle(&self, body: &str) -> Result<(), HandlerError> {
        let parsed: MessageBody = serde_json::from_str(body)?;
        match parsed {
            MessageBody::PostCreated(b) => self.handle_post_created(b).await,
        }
    }
}
